File: pkg/bus/options.go
package bus

import (
	"context"
	"log/slog"
	"time"

	"github.com/ModulationAI/openagentio/pkg/codec"
	"github.com/ModulationAI/openagentio/pkg/event"
	"github.com/ModulationAI/openagentio/pkg/middleware"
	"github.com/ModulationAI/openagentio/pkg/transport"
)

// DefaultSubjectPrefix is used when WithSubjectPrefix is not supplied.
// Override it via bus.WithSubjectPrefix("...") to coexist with legacy
// `agent.*` namespaces.
const DefaultSubjectPrefix = "acp.v1"

// EnvelopePreparer mutates an outbound envelope just before it is encoded
// onto the wire. The chain runs for every Bus.Publish / Invoke / StreamInvoke
// call, in registration order, so a later preparer observes the mutations of
// earlier ones. The canonical use case is the OpenTelemetry bridge in
// pkg/middleware/otel, which writes the active span's traceparent into
// envelope.Traceparent.
//
// Preparers must be cheap and non-blocking — they run on the publishing
// goroutine inside the hot path. Errors are not propagated; if a preparer
// needs to bail, it should leave the envelope unchanged.
type EnvelopePreparer func(ctx context.Context, e *event.Envelope)

// Options bundles every Bus-level setting.
type Options struct {
	AgentID           string
	Tenant            string
	SubjectPrefix     string
	Codec             codec.Codec
	Transport         transport.Transport
	Logger            *slog.Logger
	Middleware        []middleware.Middleware
	EnvelopePreparers []EnvelopePreparer
	DefaultTimeout    time.Duration
}

// Option mutates Options.
type Option func(*Options)

func WithAgentID(id string) Option      { return func(o *Options) { o.AgentID = id } }
func WithTenant(t string) Option        { return func(o *Options) { o.Tenant = t } }
func WithSubjectPrefix(p string) Option { return func(o *Options) { o.SubjectPrefix = p } }
func WithCodec(c codec.Codec) Option    { return func(o *Options) { o.Codec = c } }
func WithLogger(l *slog.Logger) Option  { return func(o *Options) { o.Logger = l } }

func WithTransport(t transport.Transport) Option {
	return func(o *Options) { o.Transport = t }
}

func WithMiddleware(mw ...middleware.Middleware) Option {
	return func(o *Options) { o.Middleware = append(o.Middleware, mw...) }
}

// WithEnvelopePreparer registers one or more preparers that run on every
// outbound envelope just before it is encoded. See EnvelopePreparer.
func WithEnvelopePreparer(p ...EnvelopePreparer) Option {
	return func(o *Options) { o.EnvelopePreparers = append(o.EnvelopePreparers, p...) }
}

func WithDefaultTimeout(d time.Duration) Option {
	return func(o *Options) { o.DefaultTimeout = d }
}

// --- Per-call options ---------------------------------------------------------

// SubOption configures Bus.Subscribe.
type SubOption func(*subOpts)

type subOpts struct {
	Queue string
}

// WithQueue puts the subscriber into a queue group; messages on the subject
// are load-balanced across all members of the same queue.
func WithQueue(q string) SubOption { return func(o *subOpts) { o.Queue = q } }

// InvokeOption configures Bus.Invoke / Bus.StreamInvoke.
type InvokeOption func(*invokeOpts)

type invokeOpts struct {
	Timeout     time.Duration
	IdleTimeout time.Duration
}

// WithTimeout sets the overall deadline for the invocation.
func WithTimeout(d time.Duration) InvokeOption {
	return func(o *invokeOpts) { o.Timeout = d }
}

// WithIdleTimeout sets the maximum gap between two streaming frames.
func WithIdleTimeout(d time.Duration) InvokeOption {
	return func(o *invokeOpts) { o.IdleTimeout = d }
}

// HandleOption configures HandleInvoke / HandleStream.
type HandleOption func(*handleOpts)

type handleOpts struct {
	Queue    string
	QueueSet bool
}

// WithHandleQueue places the handler in a queue group so multiple replicas
// load-balance the work.
func WithHandleQueue(q string) HandleOption {
	return func(o *handleOpts) {
		o.Queue = q
		o.QueueSet = true
	}
}

// --- Internal helpers ---------------------------------------------------------

func collectSubOpts(opts []SubOption) subOpts {
	var o subOpts
	for _, f := range opts {
		f(&o)
	}
	return o
}

func collectInvokeOpts(opts []InvokeOption) invokeOpts {
	var o invokeOpts
	for _, f := range opts {
		f(&o)
	}
	return o
}

func collectHandleOpts(opts []HandleOption) handleOpts {
	var o handleOpts
	for _, f := range opts {
		f(&o)
	}
	return o
}

File: sdk/python/src/openagentio/bus/options.py
"""Functional option pattern for Bus construction. Mirrors pkg/bus/options.go.

Provides :class:`Options`, the :data:`Option` type alias, and all ``With*``
helper functions. Per-call option types (:data:`SubOption`, :data:`InvokeOption`,
:data:`HandleOption`) are also defined here so downstream code can import them
from a single location.
"""
from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Protocol, runtime_checkable

from openagentio.bus.subjects import DEFAULT_SUBJECT_PREFIX
from openagentio.codec.json_codec import Codec, JSONCodec


# --- EnvelopePreparer ---------------------------------------------------------

EnvelopePreparer = Callable[[Any], None]


# --- Bus-level options --------------------------------------------------------

@dataclass
class Options:
    """Every Bus-level setting in one place."""

    agent_id: str = ""
    tenant: str = ""
    subject_prefix: str = DEFAULT_SUBJECT_PREFIX
    codec: Codec | None = None
    transport: Any = None  # Transport protocol instance
    logger: logging.Logger | None = None
    middleware: list[Callable] = field(default_factory=list)
    envelope_preparers: list[EnvelopePreparer] = field(default_factory=list)
    default_timeout: float = 30.0


Option = Callable[[Options], None]


def WithAgentID(id: str) -> Option:
    def apply(o: Options) -> None:
        o.agent_id = id
    return apply


def WithTransport(t: Any) -> Option:
    def apply(o: Options) -> None:
        o.transport = t
    return apply


def WithTenant(t: str) -> Option:
    def apply(o: Options) -> None:
        o.tenant = t
    return apply


def WithSubjectPrefix(p: str) -> Option:
    def apply(o: Options) -> None:
        o.subject_prefix = p
    return apply


def WithCodec(c: Codec) -> Option:
    def apply(o: Options) -> None:
        o.codec = c
    return apply


def WithLogger(l: logging.Logger) -> Option:
    def apply(o: Options) -> None:
        o.logger = l
    return apply


def WithMiddleware(*mws: Callable) -> Option:
    def apply(o: Options) -> None:
        o.middleware.extend(mws)
    return apply


def WithEnvelopePreparer(*preparers: EnvelopePreparer) -> Option:
    def apply(o: Options) -> None:
        o.envelope_preparers.extend(preparers)
    return apply


def WithDefaultTimeout(d: float) -> Option:
    def apply(o: Options) -> None:
        o.default_timeout = d
    return apply


# --- Per-call options ---------------------------------------------------------

@dataclass
class _SubOpts:
    queue: str = ""


SubOption = Callable[[_SubOpts], None]


def WithQueue(q: str) -> SubOption:
    def apply(o: _SubOpts) -> None:
        o.queue = q
    return apply


def collect_sub_opts(opts: list[SubOption]) -> _SubOpts:
    o = _SubOpts()
    for f in opts:
        f(o)
    return o


@dataclass
class _InvokeOpts:
    timeout: float | None = None
    idle_timeout: float | None = None


InvokeOption = Callable[[_InvokeOpts], None]


def WithTimeout(d: float) -> InvokeOption:
    def apply(o: _InvokeOpts) -> None:
        o.timeout = d
    return apply


def WithIdleTimeout(d: float) -> InvokeOption:
    def apply(o: _InvokeOpts) -> None:
        o.idle_timeout = d
    return apply


def collect_invoke_opts(opts: list[InvokeOption]) -> _InvokeOpts:
    o = _InvokeOpts()
    for f in opts:
        f(o)
    return o


@dataclass
class _HandleOpts:
    queue: str = ""
    queue_set: bool = False


HandleOption = Callable[[_HandleOpts], None]


def WithHandleQueue(q: str) -> HandleOption:
    def apply(o: _HandleOpts) -> None:
        o.queue = q
        o.queue_set = True
    return apply


def collect_handle_opts(opts: list[HandleOption]) -> _HandleOpts:
    o = _HandleOpts()
    for f in opts:
        f(o)
    return o